src/t_cose_encrypt_dec.c

```diff
@@ -0,0 +1,467 @@
+/*
+ * t_cose_encrypt_dec.c
+ *
+ * Copyright (c) 2022, Arm Limited. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * See BSD-3-Clause license in README.md
+ */
+#include "t_cose/t_cose_encrypt_dec.h"
+#include "t_cose_standard_constants.h"
+#include "qcbor/qcbor.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include "qcbor/qcbor_spiffy_decode.h"
+#include "t_cose_crypto.h"
+
+enum t_cose_err_t
+t_cose_encrypt_dec(struct t_cose_encrypt_dec_ctx* me,
+                   uint8_t *cose,
+                   size_t cose_len,
+                   uint8_t *detached_ciphertext,
+                   size_t detached_ciphertext_len,
+                   uint8_t *plaintext,
+                   size_t plaintext_len,
+                   size_t *plaintext_output_len)
+{
+    QCBORItem              protected_hdr;
+    QCBORItem              inner_protected_hdr;
+    UsefulBufC             nonce_cbor;
+    UsefulBufC             kid_cbor;
+    int64_t                algorithm_id = 0;
+    QCBORDecodeContext     DC, DC2;
+    QCBORItem              Item;
+    QCBORItem              Cipher;
+    QCBORError             result;
+    enum t_cose_err_t      cose_result;
+    size_t                 key_bitlen;
+    int64_t                alg = 0;
+    uint8_t               *ciphertext;
+    size_t                 ciphertext_len;
+
+    int64_t                kty;
+    int64_t                crv;
+    UsefulBufC             peer_key_x;
+    UsefulBufC             peer_key_y;
+
+    /* Temporary storge area */
+    uint8_t                tmp[50];
+    /* Temporary storge area for encrypted cek. */
+    uint8_t                tmp2[50];
+    UsefulBufC             ephemeral = {(uint8_t *) tmp, sizeof(tmp)};
+    UsefulBufC             cek_encrypted = {(uint8_t *) tmp2, sizeof(tmp2)};
+    size_t                 peer_key_buf_len = 0;
+    /*  Temporary storge area for encrypted cek. */
+    uint8_t                peer_key_buf[PSA_EXPORT_PUBLIC_KEY_MAX_SIZE] = {0x04};
+    int                    hpke_mode;
+
+    int                    ret;
+    uint8_t                add_data[20];
+    size_t                 add_data_len = sizeof(add_data);
```

Comment:
Hard coding this length is not a good long term solution because it will limit the size of headers and aad. In the short term, I think this can be a hard and documented limit, but maybe this for the long term.

- Return an error code that indicates this was too small.
- A method that lets you get the necessary size of this buffer.
- A method that lets you supply this buffer.